src/librbd/Operations.cc

```diff
@@ -1024,6 +1025,46 @@ void Operations<I>::execute_snap_unprotect(const char *snap_name,
 }
 
 template <typename I>
+int Operations<I>::set_snap_limit(const int limit) {
+  CephContext *cct = m_image_ctx.cct;
+  ldout(cct, 5) << this << " " << __func__ << ": limit=" << limit << dendl;
+
+  if (m_image_ctx.read_only) {
+    return -EROFS;
+  }
+
+  int r = m_image_ctx.state->refresh_if_required();
+  if (r < 0) {
+    return r;
+  }
+
+  {
+    RWLock::RLocker owner_lock(m_image_ctx.owner_lock);
+    C_SaferCond cond_ctx;
```

Comment:
You will have to force-acquire the exclusive lock before executing your async method.  That will ensure that if journaling is enabled, you have the journal opened.